src/ripple/basics/SlabAllocator.h

```diff
@@ -214,95 +345,102 @@ class SlabAllocator
 
-        while (slab != nullptr)
+        if (active)
         {
-            if (auto ret = slab->allocate())
+            if (auto ret = allocate(active))
                 return ret;
+        }
 
-            slab = slab->next_;
+        for (auto slab = slabs_.load(std::memory_order::acquire); slab;
+             slab = slab->next_)
+        {
+            if (auto ret = allocate(slab))
+            {
+                // This may fail, but that's OK.
+                active_.compare_exchange_weak(
+                    active, slab, std::memory_order::relaxed);
+                return ret;
+            }
         }
 
         // No slab can satisfy our request, so we attempt to allocate a new
-        // one here:
-        std::size_t size = slabSize_;
-
-        // We want to allocate the memory at a 2 MiB boundary, to make it
-        // possible to use hugepage mappings on Linux:
+        // one. We align the block at a 2 MiB boundary to allow transparent
+        // hugepage support on Linux.
         auto buf =
-            boost::alignment::aligned_alloc(megabytes(std::size_t(2)), size);
+            reinterpret_cast<std::uint8_t*>(boost::alignment::aligned_alloc(
+                megabytes(std::size_t(2)), slabSize_));
 
-        // clang-format off
-        if (!buf) [[unlikely]]
+        if (!buf)
             return nullptr;
-            // clang-format on
-
-#if BOOST_OS_LINUX
-        // When allocating large blocks, attempt to leverage Linux's
-        // transparent hugepage support. It is unclear and difficult
-        // to accurately determine if doing this impacts performance
-        // enough to justify using platform-specific tricks.
-        if (size >= megabytes(std::size_t(4)))
-            madvise(buf, size, MADV_HUGEPAGE);
-#endif
-
-        // We need to carve out a bit of memory for the slab header
-        // and then align the rest appropriately:
-        auto slabData = reinterpret_cast<void*>(
-            reinterpret_cast<std::uint8_t*>(buf) + sizeof(SlabBlock));
-        auto slabSize = size - sizeof(SlabBlock);
-
-        // This operation is essentially guaranteed not to fail but
-        // let's be careful anyways.
-        if (!boost::alignment::align(
-                itemAlignment_, itemSize_, slabData, slabSize))
+
+        // Check whether there's an existing slab with no associated buffer
+        // that we can give our newly allocated buffer to:
+        for (auto slab = slabs_.load(std::memory_order::acquire);
+             slab != nullptr;
+             slab = slab->next_)
+        {
+            if (auto ret = slab->assign_and_allocate(
+                    buf, slabSize_, itemSize_, itemCount_))
+            {
+                // This slab had available space and we used it. The block
+                // of memory we allocated is no longer needed.
+                if (ret != buf)
+                    boost::alignment::aligned_free(buf);
```

Comment:
I'm uncomfortable with this extraneous allocation and free'ing up of the `buf` 's memory.

We could split up `assign_and_allocate` into two functions -- split up at https://github.com/nbougalis/rippled/blob/a4e27672a29d39e26fedd1e074b68b71ef6b2ebd/src/ripple/basics/SlabAllocator.h#L192

This would differentiate between reusing existing buffers versus using the new buffer.

We could do a first pass over `slabs_`, if there is an existing free buffer, that can be used. Otherwise, we can proceed with the allocation for `buf`.

Do you think the benefits is worth the refactor?

Reply:
> We could split up `assign_and_allocate` into two functions -- split up at https://github.com/nbougalis/rippled/blob/a4e27672a29d39e26fedd1e074b68b71ef6b2ebd/src/ripple/basics/SlabAllocator.h#L192
> 
> This would differentiate between reusing existing buffers versus using the new buffer.
> 
> We could do a first pass over `slabs_`, if there is an existing free buffer, that can be used. Otherwise, we can proceed with the allocation for `buf`.

The iteration over `slabs_` is lock-free: that is, multiple threads can be iterating the list at the same time and, potentially, multiple threads might select the _same_ slab and try to give it a new buffer.

"Ah, don't do that!" you might say. But in order to **not** do that we'd either need to make the iteration require a lock, which will have a significant and *unnecessary* performance impact.

And locking a single slab isn't much better. First, if we lock some slab X and then allocate the memory with the slab lock held, in order to assign it, every other allocation and deallocation attempt might stall, waiting for that slab lock to be released. And _sure_, we could fix this by doing a `try_lock` and skipping a slab that's locked, but that's also additional complexity and it would only work for allocations, not deallocations.

Remember too that the system may take its sweet time to actually allocate memory. You're doing a system call, after all, and it may require a lot of work before it can be serviced.

The advantage of the current setup is that the work happens _without any locks_ and so, other memory allocation/deallocation attempts that use that slab will proceed independently at full speed.


> Do you think the benefits is worth the refactor?

I don't think that the benefits are worth the refactor; in fact, I think the net result will be a pessimization with performance implications.

Which brings us back to the opening sentence in your post:

> I'm uncomfortable with this extraneous allocation and free'ing up of the `buf` 's memory.

Why? What's making you uncomfortable?

Reply:
Thanks for the explanation about the lock-free code and its challenges, that was helpful.




> The advantage of the current setup is that the work happens _without any locks_ and so, other memory allocation/deallocation attempts that use that slab will proceed independently at full speed.
> 
To clarify, `assign_annd_allocate` function can be entered by one-thread only (https://github.com/nbougalis/rippled/blob/17f88fd151f9810693f1a5e3c41723bbd6005fbd/src/ripple/basics/SlabAllocator.h#L172)

> Why? What's making you uncomfortable?

> Remember too that the system may take its sweet time to actually allocate memory. You're doing a system call, after all, and it may require a lot of work before it can be serviced.

We are performing a memory allocation at https://github.com/nbougalis/rippled/blob/17f88fd151f9810693f1a5e3c41723bbd6005fbd/src/ripple/basics/SlabAllocator.h#L376

In one of the code paths (if a slab is found with an empty slot), that memory is not given back to the OS without any usage here: https://github.com/nbougalis/rippled/blob/17f88fd151f9810693f1a5e3c41723bbd6005fbd/src/ripple/basics/SlabAllocator.h#L394

I'm wondering if we can eliminate the unnecessary memory allocation and free system calls. Can we allocate memory only when we are absolutely sure that we will use it?

Reply:
>In one of the code paths (if a slab is found with an empty slot), that memory is not given back to the OS without any usage here:

Huh? I'm sorry but you seem confused. Let's walk through this:

We allocate some memory, and we begin iterating the slabs.

If a slab is found with a free slot, we use that slot and *release* the memory we just allocated (since it's not needed).

If a slab is found without a free slot _and_ without a memory buffer, we give it the memory we just allocated, and it gives us back a slot (it's guaranteed to have one now!)

If we run out of slabs, we create a _new_ slab and give it the memory we just allocated.

In all cases, either the memory is used, or it's properly free. There are no leaks.

>I'm wondering if we can eliminate the unnecessary memory allocation and free system calls. Can we allocate memory only when we are absolutely sure that we will use it?

No, for the reasons I explained previously. If you want to perform the allocation only after you're sure you need to, you must hold the `std::mutex` of the slab you intend to use, which is what we're trying to avoid.